### PR TITLE
feat: add tenant id to log output

### PR DESCRIPTION
**Pull request by amanji**
This PR modifies the Tenant plugin with a tenant-specific log formatter that will inject the Tenant ID into log records (replacing the default Wallet ID in the ACA-Py multitenant setup).

Tenant ID is dependent on the Wallet ID which itself is dependent on the auth token present in request headers. As a result, only tenant-authorized methods and routes will log the Tenant ID.

### CODE REVIEW

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py
Comment (jamshale) [inline plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py:129]: ** be fetched **

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/__init__.py
Comment (jamshale) [inline plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/__init__.py:36]: This needs to be here?

**Review by jamshale — DISMISSED** [review]
Looks good :+1: . Couple unimportant comments.

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/__init__.py
Comment (amanji) [reply plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/__init__.py:36]: Yeah it basically allows every log to be included

File: plugins/traction_innkeeper/pyproject.toml
Comment (loneil) [inline plugins/traction_innkeeper/pyproject.toml:12]: We should probably hold on this to align with 0.12.0 release before merging to Traction

**Review by loneil — CHANGES_REQUESTED** [review]
Just a block on merge for release management

File: plugins/traction_innkeeper/pyproject.toml
Comment (jamshale) [reply plugins/traction_innkeeper/pyproject.toml:12]: Interesting :+1:

**Comment by loneil** [comment]
@amanji should the setup on this PR be outputting tenant identifiers on the logs to standard out?

Was poking around and I don't see them in the pod
for example some snippits of creating a schema

image

File: plugins/traction_innkeeper/pyproject.toml
Comment (loneil) [reply plugins/traction_innkeeper/pyproject.toml:12]: Traction main branch (base images and innkeeper plugin dependencies) are updated to 0.12.0 now

**Comment by amanji** [comment]
The IDs only show up when using the API. It's injected as route middleware.

**Comment by loneil** [comment]
> The IDs only show up when using the API. It's injected as route middleware.

All those calls do go through the API. The Tenant UI calls the API with the token same as any Tenant LOB would.

File: plugins/pyproject.toml
Comment (loneil) [inline plugins/pyproject.toml:23]: Duplicate from merge (probably what's causing build failure)

File: plugins/traction_innkeeper/pyproject.toml
Comment (loneil) [inline plugins/traction_innkeeper/pyproject.toml:17]: Duplicate, likely from merge (probably what's causing build failure)

File: plugins/pyproject.toml
Comment (amanji) [reply plugins/pyproject.toml]: Removed

File: plugins/traction_innkeeper/pyproject.toml
Comment (amanji) [reply plugins/traction_innkeeper/pyproject.toml:17]: If it fails, I'll rerun to update the lock file or revert

**Comment by amanji** [comment]
Not sure why the IDs are not displaying. At minimum the Wallet ID should show up in the logs.

**Review by loneil — CHANGES_REQUESTED** [review]
FYI this is rebased to aca-py 0.12.1 now. Still not seeing wallet or tenant id in standard out logs on it.

**Comment by esune** [comment]
> FYI this is rebased to aca-py 0.12.1 now. Still not seeing wallet or tenant id in standard out logs on it.

@amanji can you please take a look?

**Comment by amanji** [comment]
Resolves: #1037

**Review by loneil — CHANGES_REQUESTED** [review]
Not sure where it's coming from (see below) but I'm getting 502 responses on this PR, or running locally with docker compose for any calls using a token. Either Innkeeper or Tenant.

image

The call to get the token works but using that token in any api endpoint isn't returning.

Wierdly not seeing any errors in the log, just the calls to get profile for token

image

**Review by loneil — DISMISSED** [review]
Calls are working now, trying out I've noticed something though.
If I make a call for a specific tenant (lets call it **Tenant 1**: 9ed537b7-d91a-4c04-be97-8423b63d3ebd ID in these examples), then any unauthenticated call _after_ that will maintain the Tenant ID from that last call.

For example. I do something with 9ed537b7-d91a-4c04-be97-8423b63d3ebd, then ready and live logs after that will look like they're made by the tenant 
```
2024-05-09 20:59:13,545 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/aiohttp/web_log.py:211 10.97.200.59 [09/May/2024:20:59:13 +0000] "GET /status/ready HTTP/1.0" 200 172 "-" "kube-probe/1.26"
2024-05-09 20:59:18,545 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/aiohttp/web_log.py:211 10.97.200.59 [09/May/2024:20:59:18 +0000] "GET /status/ready HTTP/1.0" 200 172 "-" "kube-probe/1.26"
```

Or another Tenant 2 goes and gets a token, since those calls are prior to auth, the Tenant 1 token shows in those logs
```
2024-05-09 21:56:04,387 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/aiohttp/web_log.py:211 10.97.200.59 [09/May/2024:21:56:04 +0000] "GET /status/ready HTTP/1.0" 200 172 "-" "kube-probe/1.26"
2024-05-09 21:56:04,393 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/multitenant_provider/v1_0/manager.py:191 < create_auth_token
2024-05-09 21:56:04,393 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/multitenant_provider/v1_0/manager.py:328 < create_auth_token
2024-05-09 21:56:04,394 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/aiohttp/web_log.py:211 10.97.200.59 [09/May/2024:21:56:00 +0000] "POST /multitenancy/wallet/aa5a2a05-b8dd-4237-96fb-5f98fad0d2a3/token HTTP/1.0" 200 564 "https://pr-1065-traction-tenant-ui-dev.apps.silver.devops.gov.bc.ca/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
2024-05-09 21:56:04,551 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/multitenant_provider/v1_0/manager.py:335 > get_profile_for_token
2024-05-09 21:56:04,551 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd INFO /home/aries/.venv/lib/python3.9/site-packages/multitenant_provider/v1_0/manager.py:198 > get_profile_for_token
```


Or also any other fetch call (from someone else or some other system or anything) fails because the password or user is is wrong, then it looks like an attempt to log in on Tenant 1.
```
2024-05-09 22:00:23,328 TENANT: 9ed537b7-d91a-4c04-be97-8423b63d3ebd ERROR /home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/admin/server.py:198 Handler error with exception: Record not found: wallet_record/123.
... trace here ...
aiohttp.web_exceptions.HTTPNotFound: Record not found: wallet_record/123.
```

So I think some unauthenticated call handling needs to be in or something?

**Comment by amanji** [comment]
Good catch. I'm only setting the tenant_id if one is found. I should be resetting it.

**Comment by loneil** [comment]
@amanji Things were working away but I saw the pod crashing out a couple times (thought it might be a resource thing and OCP was being suspiciously slow) but then digging I've seen a couple pods end with

```
Traceback (most recent call last):
File "/home/aries/.venv/bin/aca-py", line 8, in <module>
sys.exit(script_main())
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 69, in script_main
main(sys.argv)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 75, in main
run(args)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 64, in run
run_command(command, args)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/commands/__init__.py", line 38, in run_command
module.execute(argv)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/commands/start.py", line 63, in execute
run_loop(start_app(conductor), shutdown_app(conductor))
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/commands/start.py", line 101, in run_loop
loop.run_forever()
File "/home/aries/.venv/lib/python3.9/site-packages/nest_asyncio.py", line 82, in run_forever
self._run_once()
File "/home/aries/.venv/lib/python3.9/site-packages/nest_asyncio.py", line 129, in _run_once
handle._run()
File "/usr/local/lib/python3.9/asyncio/events.py", line 94, in _run
self._loop.call_exception_handler(context)
File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1786, in call_exception_handler
logger.error('Exception in default exception handler',
File "/usr/local/lib/python3.9/logging/__init__.py", line 1475, in error
self._log(ERROR, msg, args, **kwargs)
File "/usr/local/lib/python3.9/logging/__init__.py", line 1587, in _log
record = self.makeRecord(self.name, level, fn, lno, msg, args,
File "/usr/local/lib/python3.9/logging/__init__.py", line 1556, in makeRecord
rv = _logRecordFactory(name, level, fn, lno, msg, args, exc_info, func,
File "/home/aries/traction_plugins/traction_innkeeper/v1_0/tenant/__init__.py", line 53, in new_log_record_factory
record = prev_log_record_factory(*args, **kwargs)
File "/home/aries/traction_plugins/traction_innkeeper/v1_0/tenant/__init__.py", line 53, in new_log_record_factory
record = prev_log_record_factory(*args, **kwargs)
File "/home/aries/traction_plugins/traction_innkeeper/v1_0/tenant/__init__.py", line 53, in new_log_record_factory
record = prev_log_record_factory(*args, **kwargs)
[Previous line repeated 983 more times]
RecursionError: maximum recursion depth exceeded
```

Can't reproduce so far but `prev_log_record_factory` is something added here right? Any thoughts?

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py
Comment (jamshale) [inline plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py:102]: Not sure if this gets cached or not? Because it's middleware it would be ideal if it was. It's just this traction plugin though so not a big deal.

**Comment by jamshale** [comment]
I'm a bit unsure where this is at? I'll approve it though if you think it's good @amanji ?

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py
Comment (esune) [inline plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py:106]: Running this on every request that goes through the middleware will cause the custom factory to create a new object to handle that specific request, that is used once and then replaced by a - identical in functionality - instance. if we were running out of resources in the PR pods this might be why.

Is there a way to set-up the log factory only *once* when the services start, rather than re-creating it for every single request? A singleton pattern comes to mind (i.e.: https://python-patterns.guide/gang-of-four/singleton), but there may be a better "pythonic" way to do this.

**Review by esune — CHANGES_REQUESTED** [review]

**Comment by esune** [comment]
> I'm a bit unsure where this is at? I'll approve it though if you think it's good @amanji ?

I think we need some more tweaks - see review comments. I'll run through the deployed PR later to check the output (haven't checked that yet), but the repeated instantiation of a log factory class is what I am not liking too much currently.

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/__init__.py
Comment (amanji) [reply plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/__init__.py:35]: This function is called as part of the logging and needs to be named accordingly

File: plugins/.devcontainer/configs/devcontainer.yml
Comment (amanji) [reply plugins/.devcontainer/configs/devcontainer.yml:61]: This is just for the devcontainers. It was causing issues at startup so I just commented it out

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py
Comment (amanji) [reply plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py:106]: Have you seen any pods crashing since this most recent change? I'm not sure how else to inject the `tenant_id` another way.

**Comment by amanji** [comment]
I've referenced the documentation on how the logging factory works

https://docs.python.org/3/howto/logging-cookbook.html#customizing-logrecord

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py
Comment (esune) [reply plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py:106]: I have not (yet) put the PR instance through some load - will try later and report back.

**Comment by esune** [comment]
> I've referenced the documentation on how the logging factory works
> 
> https://docs.python.org/3/howto/logging-cookbook.html#customizing-logrecord
> 
> I will also take another stab at filtering:
> 
> https://docs.python.org/3/library/logging.html#filter-objects

Thanks! I think the overall approach is correct, although I am wondering what would happen if the innkeeper plugin wasn't installed/run last in the plugin stack. This is something we can ignore though, since we are in control of how things are setup for Traction and can specify the plugin to be the last one to be loaded in the startup parameters (I believe).

My main concern is mostly with creating/setting a new log factory for every request, based on its creation being part of the middleware: ideally, this would be something that is set up front by the plugin and used from there on without having to re-create the factory class again.

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py
Comment (esune) [reply plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py:102]: I think this is basically the same as my comment. Will do a performance assessment in `dev` and see if we need to address this at all.

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py
Comment (esune) [reply plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py:106]: Will do further performance assessment in dev and see if/how to tackle this.

**Review by esune — APPROVED** [review]

**Pull request merged**
